### PR TITLE
Implement sequential[B](tasks: Seq[Initialize[Task[B]]]) and remove useless coment outs

### PR DESCRIPTION
**Pull request by 3tty0n**
This implements `sequental[B](tasks: Seq[Initialize[Task[B]]])` method in `TaskSequential.scala` and remove useless comment outs.

The return value of `sequential[B](tasks: Seq[Initialize[Task[B]]])` is the last one of `tasks`.

Before:
```scala
lazy val tasks = Seq(taskA, taskB, taskC)
test in Test := Def.sequential(tasks.init, tasks.last).value // a little confusing
```

After:
```scala
lazy val tasks = Seq(taskA, taskB, taskC)
test in Test := Def.sequential(tasks).value // looks nice
```

### CODE REVIEW

File: notes/1.0.0/fix-task-sequential.markdown [hunk 1]
```diff
@@ -0,0 +1,5 @@
+### Impprovements
```
Comment (jvican) [inline notes/1.0.0/fix-task-sequential.markdown]: Extra 'p' here :smile:!

**Review by jvican — APPROVED** [review]
Hey @3tty0n, tihs looks great! I've always wondered why it wasn't easier to do this. LGTM.
  
Thanks for taking the time to report the issue and address it in a PR. Small changes matter :smile:.

File: notes/1.0.0/fix-task-sequential.markdown (hunk 1 above)
Comment (3tty0n) [reply notes/1.0.0/fix-task-sequential.markdown]: Thank you for your quick reply!
I'll fix it as soon as possible.

**Comment by 3tty0n** [comment]
I fixed the typo in the notes, but the CI was failed (even though the previous CI test was succeessful). Should I update the branch?

EDIT:
I updated this branch.

**Comment by dwijnand** [comment]
I don't like the addition of this overload because it could fail Nil. Put differently: it uses init and last which aren't partial functions.

[edit: ... which are* partial functions]

**Comment by jvican** [comment]
That makes sense. Can we special case for list of only one element and return the task itself instead of calling the underlying sequential?

**Comment by jvican** [comment]
*for one and zero elements.

**Comment by 3tty0n** [comment]
I agree with this idea.
> Can we special case for list of only one element and return the task itself instead of calling the underlying sequential?

**Comment by 3tty0n** [comment]
I will implement and push it when I go back home.

**Comment by 3tty0n** [comment]
I have one problem.
How `sequential[B](tasks: Seq[Initialize[Task[B]]])` behave when `tasks` is `Nil` ?
It seems good to do nothing, but `Task[Unit]` is not inferred to `Task[B]]`.

---

```scala
def sequential[B](tasks: Seq[Initialize[Task[B]]]): Initialize[Task[B]] =
  tasks.toList match {
     case Nil => ???
     case x :: Nil => Def.task { x.value }
     case _ =>
       sequential(tasks.init.map(unitTask), tasks.last)
   }
```

**Comment by jvican** [comment]
@3tty0n Given the tradition of sbt of throwing exceptions all over the place, I would just throw an exception if the list is zero, and document it with Scaladoc.

**Comment by ShaneDelmore** [comment]
What about using a NonEmptySeq?

**Comment by 3tty0n** [comment]
@jvican Thanks. I will try it as you told me.

@ShaneDelmore NonEmptySeq is built-in data structure? If not, I think using NonEmptySeq is not recommended.

**Comment by dwijnand** [comment]
Let's not add partial methods to the API.

**Comment by jvican** [comment]
The full of sbt API, both public and internal, is full of them. I see why we shouldn't, honestly.

**Comment by jvican** [comment]
Okay, I agree this is far than ideal. What if we just return an empty task if the argument of sequential is empty? Then we're total. We should document it well in the public API. Are you happier with this @dwijnand?

**Comment by dwijnand** [comment]
What's an empty task? Given 0 tasks of `B` how are you going to return 1 task of `B`?

**Comment by jvican** [comment]
Got confused by the `unitTask` in `SequentialTask`. I see two ways to move this forward:
  
1. Use https://github.com/tarao/nonempty-scala. Instead of `List(1,2,3)` we can do `NonEmpty(1,2,3)` and ensure that the collection is non empty. Pretty neat. The idea would be to port it to sbt/sbt.
2. Require `::`. The major disadvantage with this approach is that `1::Nil` returns `List` not `::`. So users of the `sequential` function asking for a list would need to patmat on the list before calling the method.
  
IMO, the first one is the best. I can see `NonEmpty` being useful in other places of the sbt API. I think it's a good idea to port it.

**Comment by jvican** [comment]
@3tty0n I think the best solution is to port https://github.com/tarao/nonempty-scala to sbt. It would be a very meaningful addition because it would allow us to enforce better the contract of the APIs and ensuring that collections are non-empty by design. Then, we can use it in the following way:

```scala
lazy val tasks = NonEmpty(taskA, taskB, taskC)
test in Test := Def.sequential(tasks).value
```
  
/cc @eed3si9n

**Comment by 3tty0n** [comment]
@jvican Thanks. I think porting and using `NonEmpty` is good way to realize the idea of this PR.

**Comment by eed3si9n** [comment]
Looking at this now, I kind of feel like adding NonEmptyList just for this sounds like a maintenance burden. Is `Def.sequential(tasks.init, tasks.last).value` that bad?

**Comment by eed3si9n** [comment]
Closing this in favor of https://github.com/sbt/sbt/pull/4369

**Pull request closed without merging**